File: 251001/Vasilenko/publisher/src/main/java/org/ex/distributed_computing/controller/WriterController.java
package org.ex.distributed_computing.controller;

import jakarta.validation.Valid;

import java.util.List;
import lombok.RequiredArgsConstructor;
import org.ex.distributed_computing.dto.request.WriterRequestDTO;
import org.ex.distributed_computing.dto.response.WriterResponseDTO;
import org.ex.distributed_computing.service.WriterService;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequiredArgsConstructor
@RequestMapping("/writers")
public class WriterController {

  private final WriterService writerService;

  @GetMapping
  public List<WriterResponseDTO> getAllWriters() {
    return writerService.getAllWriters();
  }

  @GetMapping("/{id}")
  public ResponseEntity<WriterResponseDTO> getWriterById(@PathVariable Long id) {
    return ResponseEntity.ok(writerService.getWriterById(id));
  }

  @PostMapping
  public ResponseEntity<WriterResponseDTO> createWriter(@Valid @RequestBody WriterRequestDTO requestDTO) {
    var author = writerService.createWriter(requestDTO);
    return ResponseEntity.status(HttpStatus.CREATED.value()).body(author);
  }

  @PutMapping()
  public ResponseEntity<WriterResponseDTO> updateWriter(@Valid @RequestBody WriterRequestDTO dto) {
    return ResponseEntity.ok(writerService.updateWriter(dto));
  }

  @DeleteMapping("/{id}")
  public ResponseEntity<Void> deleteWriter(@PathVariable Long id) {
    writerService.deleteWriter(id);
    return ResponseEntity.noContent().build();
  }
}

File: 251003/Pankratiev/rest/src/main/kotlin/bsuir/dc/rest/controller/WriterController.kt
package bsuir.dc.rest.controller

import bsuir.dc.rest.dto.from.WriterFrom
import bsuir.dc.rest.dto.to.WriterTo
import bsuir.dc.rest.service.WriterService
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/v1.0/writers")
class WriterController(
    private val writerService: WriterService
) {

    @PostMapping
    fun createWriter(@RequestBody @Valid writerFrom: WriterFrom): ResponseEntity<WriterTo> {
        val createdWriter = writerService.createWriter(writerFrom)
        return ResponseEntity(createdWriter, HttpStatus.CREATED)
    }

    @GetMapping("/{id}")
    fun getWriterById(@PathVariable id: Long): ResponseEntity<WriterTo> {
        val writer = writerService.getWriterById(id)
        return ResponseEntity.ok(writer)
    }

    @GetMapping
    fun getAllWriters(): ResponseEntity<List<WriterTo>> {
        val writers = writerService.getAllWriters()
        return ResponseEntity.ok(writers)
    }

    @PutMapping()
    fun updateWriter(@RequestBody @Valid writerFrom: WriterFrom): ResponseEntity<WriterTo> {
        val updatedWriter = writerService.updateWriter(writerFrom.id, writerFrom)
        return ResponseEntity.ok(updatedWriter)
    }

    @DeleteMapping("/{id}")
    fun deleteWriter(@PathVariable id: Long): ResponseEntity<Void> {
        writerService.deleteWriter(id)
        return ResponseEntity.noContent().build()
    }

    @GetMapping("/by-issue/{issueId}")
    fun getWriterByIssueId(@PathVariable issueId: Long): ResponseEntity<WriterTo> {
        val writer = writerService.getWriterByIssueId(issueId)
        return ResponseEntity.ok(writer)
    }
}
